### PR TITLE
Wrote initial draft of design document.

### PR DESCRIPTION
**Pull request by Zagitta**
A brief design/approach document.

### CODE REVIEW

File: Approach.md
Comment (knisbet) [inline Approach.md]: To demo an unauthorized client may not need a CA and can just be self signed... if that helps simplify it at all. But nothing wrong with a separate CA, just wanted to mention it as an option.

File: Approach.md
Comment (knisbet) [inline Approach.md]: > The public key of a cert will be used as the client id despite the drawback that a single client would have to share the same cert across multiple servers against best practices. 

Sorry, I'm not sure I follow what the drawback / best practice being violated is here?

File: Approach.md
Comment (knisbet) [inline Approach.md]: Could you please point me towards a system that implements a model like described here. I'm not totally sure I understand / follow what the advice would be here for a production environment.

File: Approach.md
Comment (knisbet) [inline Approach.md]: If I'm following this correctly, I wanted to mention that you may need to be careful that you're not blocking threads that could impact the service with this approach of holding a lock while an external operation to complete.

File: Approach.md
Comment (knisbet) [inline Approach.md]: What about other aspects of security, like the isolation of the containerization? Will any protections be taken against container escape for example?

**Review by knisbet — COMMENTED** [review]
One thing we often ask for is to include the proto spec in the design, so we have an early opportunity to add any comments or thoughts to the API before spending too much time on implementation.

File: Approach.md
Comment (Zagitta) [reply Approach.md]: My understanding of best practices are to keep private keys local and never transfer them between systems as there are risk this might be done in an insecure way.
In a similar vein it's best practices regarding ssh keys for development where it's recommended to generate a unique ssh key for each system you expect to push code from.
This helps with damage control should one of the systems be compromised as you'll only need to revoke one key and investigate commits made with that key.

As such the drawback here is that this corner cutting forces the client to share the same private key across multiple servers in order to access the same group of tasks from multiple servers since different key pairs will see a set of tasks unique to that set of keys.

See my answer to your below question for what I'd recommend for an actual production system.

File: Approach.md
Comment (Zagitta) [reply Approach.md]: The indirection I mention here could be implemented in two ways from the top of my head. 
One would be database backed where an authenticated request has it's public key looked up in a table that maps between public keys and client_id which is then used to return the appropriate set of tasks. This is similar to how github maps a multiple ssh keys to one github account.
Another approach could be to use **our** client CA to sign a subclient CA with which the customer could sign a certificate for each of their servers and then include the subclient CA in the certificate chain used for authentication. I'm not sure there are any production systems using this model as it's rather cumbersome for the customer to manage their own CA.

File: Approach.md
Comment (Zagitta) [reply Approach.md]: I intend to write the service as a fully async implementation and rely on the tokio async RWLock which should allow the async runtime to continue processing other requests while waiting for the stop task to finish.
But yes it's a good point there's some complexity here that needs to be handled carefully.

File: Approach.md
Comment (Zagitta) [reply Approach.md]: I will be relying on the `clone(2)` and `pivot_root(2)` to isolate the containers so barring any bugs in my implementation I'll be relying on the security of the kernel.

However briefly mentioned in the test section I'll write some basic unit tests that verifies a containerized task is unable to see the host system PIDs, make network connections and write to unauthorized paths.

More advanced container escape attacks are outside my expertise and will be considered out of scope. 
However I'd be more than happy learn about such attacks :-)

File: Approach.md
Comment (Zagitta) [reply Approach.md]: Good point, no need to introduce an extra CA. I'll fix this.

**Comment by Zagitta** [comment]
> One thing we often ask for is to include the proto spec in the design, so we have an early opportunity to add any comments or thoughts to the API before spending too much time on implementation.

Yes that a good point and I did think about this and it was my plan to include it along with some minor coding stuff in PR #1 per the timeline section.
However I already have a draft for the API spec so I'll complete it later and add it to this PR as requested.

File: Approach.md
Comment (xacrimon) [reply Approach.md]: Might also want to look into the tokio semaphore. Rwlocks are meant for protecting data while the semaphores are general concurrency control for external things.

File: Approach.md
Comment (xacrimon) [reply Approach.md]: I think we generally ask to use cgroups in order to provide robust namespacing and resource limitation as specified by the challenge.

File: Approach.md
Comment (xacrimon) [inline Approach.md]: Missing a CLI/UX section

File: Approach.md
Comment (Zagitta) [reply Approach.md]: As far as I'm aware cgroups only provides resource utilization constraints. Namespaces provide isolation and can be activated with the `unshare(2)` and `clone(2)` syscalls.
Since I'm aiming for level 5 with both utilization limits and isolation I'll be taking advantage of both cgroups and namespaces.

File: Approach.md
Comment (Zagitta) [reply Approach.md]: Good point, added!

**Comment by Zagitta** [comment]
@knisbet I've added the protbuff API as requested.

File: Approach.md
Comment (Zagitta) [reply Approach.md]: Protecting the data is exactly what I'm going for. The idea is to write something along the lines of a `ChildHandle` with a `async fn kill(&mut self) -> Result<_,_>` such that there can only ever be a single task attempting to kill the child, all others will be blocked on `let child = rw_lock.write().await;` and once they gain write access the previous task will already have marked it as killed and prevent any data races.
Of course that has the drawback that multiple stop requests to the same child will queue up and maybe even timeout on the gRPC end but I think that's an ok tradeoff for this challenge.
In a real production system you'd make a proper error for this edgecase.

File: rrocker-lib/protos/api.proto
Comment (knisbet) [inline rrocker-lib/protos/api.proto]: By using a string to represent output, what happens if a particular line of output is very long? Or contains bytes that are not valid as a string?

File: Approach.md
Comment (knisbet) [reply Approach.md]: Ok, I think I follow the line of argument. This is mostly about avoiding the copying of private keys to different hosts acting as a client that may want to present the same identity to the service. 

So the way this is often approached in mTLS is to not use the public key as the identity, but the subject field of the certificate. So using kubernetes as an example, you would issue certificates with something like `Subject: CN=kevin, O=cloud-dev O=security`. Kubernetes when it receives a certificate like this, will take the subject and read the CN as the user and the O fields as any groups. So kubernetes will interpret that Subject as user=kevin and groups=cloud-dev,security when it applies RBAC on what API requests I'm allowed to make based on both groups and principal. 

To uniquely identify an issued certificate, would be based on the Serial Number. So if you want to selectively disable a certificate, you would find the serial and add it to a revocation list, if you're using standard revocation, which has it's faults. For user identity, another common approach is to issue short lived certificates. So if the certificate is only valid for one hour, the expiry takes care of cancelling credentials that may leak.

File: rrocker-lib/protos/api.proto
Comment (knisbet) [inline rrocker-lib/protos/api.proto]: I wanted to mention incase you aren't aware, but gRPC does have it's own error mechanism and a number of predefined codes/results that are outside the message definition. I believe in tonic the codes are covered here: https://docs.rs/tonic/0.5.0/tonic/enum.Code.html

File: rrocker-lib/protos/api.proto
Comment (Zagitta) [reply rrocker-lib/protos/api.proto]: I go into this a bit in the approach document under the rrockerd section:
> Dealing with tty, ptty, terminals and what not is entirely out of scope and as such any program output will be streamed as the raw bytes written to the stdout and stderr pipes.
Furthermore the output will be chunked on newlines and any lines longer than max gRPC packet size will be truncated.

So any line longer than >4MB ([default max recv value](https://github.com/grpc/grpc/blob/master/include/grpc/impl/codegen/grpc_types.h#L495)) will be truncated.
But yes you're right, I didn't consider that gRPC strings must be valid utf8.

This area is a huge can of worm. For example there are no guarantees a program even writes valid utf8 and it could just spit raw binary out in which case even attempting to chunk it on newlines is futile. 
A pathological case would happen with something as simple as `cat some_big_bin_file`.
I chose newline chunking as that was the simplest and in my opinion the the stream command is meant to be used as log viewer.
As such it made sense to do it the simple way with a lossy utf8 conversion and dropping unprintable characters on the daemon side.
One could also choose to encode each byte as a separate message but that's clearly not scalable. 
Another solution is to then chunk on some max length like 1024 bytes of output but that complicates the problem because now you also need to have a timeout for sending partial chunks which introduces latency.

In a production system I'd probably make some kind of hybrid system that attempts to detect whether the program output is text or binary and then switches between chunking on newlines or number of bytes.

So for this challenge I'll cut a big corner by assuming the output is meant to be human readable and discard invalid utf8.
Of course that means you couldn't do something like `rrocker-cli stream $(rrocker-cli start /bin/cat /dev/sda) > backup.bin` but I think that's a worthwhile tradeoff for this challenge.

File: rrocker-lib/protos/api.proto
Comment (Zagitta) [reply rrocker-lib/protos/api.proto]: You're right, I somehow missed that entirely. 
That'll make things quite a bit cleaner, I'll fix it right away.

File: Approach.md
Comment (Zagitta) [reply Approach.md]: Yes that's indeed a much better approach than what I came up with. I've updated the design doc to reflect this change. Thanks!

File: rrocker-lib/protos/api.proto
Comment (russjones) [inline rrocker-lib/protos/api.proto]: Why do you return output or state here? You have another endpoint for state right?

File: Approach.md
Comment (russjones) [inline Approach.md:40]: Can describe how you plan to handle streaming?

File: rrocker-lib/protos/api.proto
Comment (Zagitta) [reply rrocker-lib/protos/api.proto]: Yes there's another endpoint solely for state.
The state here will only be sent as the last message to indicate why a stream ended rather than requiring an extra round trip to query the task state.

File: Approach.md
Comment (Zagitta) [reply Approach.md:40]: Done! Let me know if you have any further questions.

File: Approach.md
Comment (russjones) [inline Approach.md:44]: I don't think this will work, in-fact I recently fixed a similar issue.

https://github.com/gravitational/teleport/pull/7449

The problem is unless you have something reading from the pipe at the same time, you can only write so much until it will block.

File: rrocker-lib/protos/api.proto
Comment (russjones) [reply rrocker-lib/protos/api.proto]: Does the Rust GRPC library not return something like `io.EOF` that the Go library does to indicate the stream ended?

**Review by knisbet — APPROVED** [review]
looks good on my end when the last couple comments from russjones are addressed.

File: Approach.md
Comment (Zagitta) [reply Approach.md:44]: So having looked at the issue and the surrounding code I think I understand the issue you fixed. 
It seems teleport itself ended up hanging since you write the command along with it's parameters over a pipe which is then read from the child process that executes the command after being placed in a cgroup.
The write to this pipe then ends up blocking with long commands due to the max size of 65k and your fix was to essentially move this write into a separate task.

As such I'll only ever read the stdout/stderr pipes from the daemon side and I'll spawn an async task for each process that gets executed such that it doesn't block anything.
One way this could cause problems is if the daemon reader task can't keep up with the output rate of the spawned process and as such could end up blocking it. 
However that's the same behavior you'd experience if you simply piped the command to a file from your terminal aka: `./spam_prog > log.txt`

My approach is to use `clone(2)` with the right namespace flags to create a forked child that becomes the init process. 
Since this new process gets CoW access to the parent daemon process's address space the arguments can be accessed directly in memory and don't have to go through a pipe.

**Review by xacrimon — APPROVED** [review]

File: rrocker-lib/protos/api.proto
Comment (Zagitta) [reply rrocker-lib/protos/api.proto]: The EOF part is handled automatically when the write handle of the gRPC stream gets dropped so the idea here is *not* to mark the end of the stream it's merely a convenience thing.
But I suppose there's a good argument to be made that it introduces extra handling on the client side for every log message received so I'll remove it.

File: Approach.md
Comment (fspmarshall) [inline Approach.md]: The streaming endpoint should actually start from the beginning of the process's output, even if the job has already exited.  No cleanup mechanism is required though.  Its OK for server memory to grow indefinitely as mode jobs are run.  Its more important that we have consistent and compete output.

File: Approach.md
Comment (russjones) [reply Approach.md:44]: Maybe I am not understanding things, but I think we are talking about two different things. The problem I see with the pipe approach is that:

1) You can't buffer a lot in the pipe.
2) Since each client has to be able to read from the start, not sure how the pipe is useful here.

Comment (Zagitta) [reply Approach.md:44]: The pipe will continuously be read out by one dedicated task per spawned process, so each pipe is only read from one place. This along with the pipe being marked as blocking by default means the spawned process will get blocked should the reader task be unable to keep up so the pipe buffer size shouldn't cause any issues.
@fspmarshall sorta touched on your second point below, it wasn't clear to me it was a requirement to get past output as well but I've updated the design doc and described how I'll implement this.
Long story short the pipe only servers as a tool to grab the programs output, each line will go through some buffering and synchronization in order to stream it out over gRPC.

File: Approach.md
Comment (Zagitta) [reply Approach.md]: It wasn't clear to me from the specification that it's a requirement to output the full output and not just future output but yes you're right it makes more sense.
Please have a look at the newest commit, I've described how I'll implement full output streaming.

**Review by fspmarshall — APPROVED** [review]

**Pull request merged**
